File: src/main/java/com/est/curdsample/domain/Task.java
package com.est.curdsample.domain;

import com.est.curdsample.dto.TaskDto;
import com.est.curdsample.util.TimeFormatter;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import java.time.LocalDate;
import java.time.LocalDateTime;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Entity
@Table(name = "tasks")
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Task {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Setter
    @Column(nullable = false, unique = true)
    private String code;

    private String title;
    private String description;

    private Integer priority;

    private boolean completeStatus = false;

    private LocalDate startTime;
    private LocalDate endTime;

    private LocalDateTime createdAt = LocalDateTime.now();
    private LocalDateTime updatedAt = LocalDateTime.now();

    public void updateCheck(){
        this.completeStatus = !this.completeStatus;
        this.updatedAt = LocalDateTime.now();
    }

    public void update(TaskDto taskDto) {

        this.title = taskDto.getTitle();
        this.description = taskDto.getDescription();
        this.priority = taskDto.getPriority();
        this.completeStatus = taskDto.isCompleteStatus();

        this.startTime = TimeFormatter.convertToLocalDate(taskDto.getStartTime());
        this.endTime = TimeFormatter.convertToLocalDate(taskDto.getEndTime());
    }

    public static Task of(TaskDto taskDto) {
        Task task = new Task();

        task.code = taskDto.getCode();
        task.title = taskDto.getTitle();
        task.description = taskDto.getDescription();
        task.completeStatus = taskDto.isCompleteStatus();
        task.startTime = TimeFormatter.convertToLocalDate(taskDto.getStartTime());
        task.endTime = TimeFormatter.convertToLocalDate(taskDto.getEndTime());

        return task;
    }

}

File: src/main/java/com/est/curdsample/domain/Task.kt
package com.est.curdsample.domain

import com.est.curdsample.dto.TaskDescription
import com.est.curdsample.dto.TaskDto
import com.est.curdsample.util.convertToLocalDate
import com.est.curdsample.util.convertToString
import jakarta.persistence.*
import lombok.Setter
import java.time.LocalDate
import java.time.LocalDateTime

@Entity
@Table(name = "tasks")
class Task(

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,

    @Setter
    @Column(nullable = false, unique = true)
    var code: String,

    var title: String,

    var description: String,

    var priority: Int,

    var completeStatus: Boolean,

    var startTime: LocalDate,

    var endTime: LocalDate,

    val createdAt: LocalDateTime = LocalDateTime.now(),

    var updatedAt: LocalDateTime = LocalDateTime.now()

) {

    fun updateCheck() {
        this.completeStatus = !this.completeStatus
        this.updatedAt = LocalDateTime.now()
    }

    fun update(taskDto: TaskDto) {
        this.title = taskDto.title
        this.description = taskDto.description
        this.priority = taskDto.priority
        this.completeStatus = taskDto.completeStatus

        this.startTime = convertToLocalDate(taskDto.startTime)
        this.endTime = convertToLocalDate(taskDto.endTime)
    }

}

fun Task.toDto(): TaskDto {
    return TaskDto(
        code = this.code,
        title = this.title,
        description = this.description,
        priority = this.priority,
        completeStatus = this.completeStatus,
        startTime = convertToString(this.startTime),
        endTime = convertToString(this.endTime),
    )
}

fun Task.toDescription(): TaskDescription {
    return TaskDescription(
        code = this.code,
        title = this.title,
        description = this.description,
        priority = this.priority,
        completeStatus = this.completeStatus,
        startDate = convertToString(this.startTime),
        dueDate = convertToString(this.endTime),
        createdAt = convertToString(this.createdAt),
        updatedAt = convertToString(this.updatedAt)
    )
}